### PR TITLE
Support Multi-Value QueryParam/Headers/Cookies

### PR DESCRIPTION
**Pull request by SentryMan**
Got a request from discord to add this.

- Creates a new Collection Type Handler specifically for those param types
- adds set, list, and a collection version of `withDefault` method to Param Converter
- adds new Platform Adapter method to handle multi-value parameters
- adds new Platform Adapter method to handle map parameters
- Can do `Map<String, List<String>>` for certain params.

### CODE REVIEW

File: http-api/src/main/java/io/avaje/http/api/PathTypeConversion.java [hunk 1]
```diff
@@ -22,6 +28,10 @@ public static String withDefault(String value, String defaultValue) {
     return value != null ? value : defaultValue;
   }
 
+  public static List<String> withDefault(List<String> value, String defaultValue) {
```
Comment (LoonyRules) [inline http-api/src/main/java/io/avaje/http/api/PathTypeConversion.java:31]: Just allow us to define a `List<String> defaultValue` parameter instead? Some defaults may include multiple entries.

File: http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java [hunk 2]
```diff
@@ -56,6 +70,46 @@ public class ElementReader {
...
+      final var mainTypeEnum =
+          typeOp
+              .flatMap(t -> Optional.ofNullable(ctx.typeElement(t.mainType())))
+              .map(TypeElement::getKind)
+              .filter(ElementKind.ENUM::equals)
+              .isPresent();
+
+      final var isCollection =
+          typeOp
+              .filter(t -> t.isGeneric() && !t.mainType().startsWith("java.util.Map"))
```
Comment (LoonyRules) [inline http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java:89]: `Optional<>` params are gonna haunt you with this.

File: http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java [hunk 3]
```diff
@@ -56,6 +70,46 @@ public class ElementReader {
...
+              .isPresent();
+
+      if (mainTypeEnum) {
+        return TypeMap.enumParamHandler(type);
+      } else if (isCollection) {
+        this.isParamCollection = true;
+        if (paramType == ParamType.FORMPARAM) {
+          throw new IllegalStateException("You can't have a single Form Parameter be a list");
```
Comment (LoonyRules) [inline http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java:97]: Code example for what this is trying to catch? Doesn't sound right to me?

File: http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java (hunk 2 above)
Comment (LoonyRules) [reply http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java:89]: Also, what if my BeanParam has a Map<String, String> field in it? Retrofit2 supports all of this stuff and can be deserialized by general json deserialization as any "BeanParam" would be treated as a Body request (GET/POST/PATCH/PUT). Could we just do the same?

Comment (SentryMan) [reply http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java:89]: Oh we already support maps

File: http-api/src/main/java/io/avaje/http/api/PathTypeConversion.java (hunk 1 above)
Comment (SentryMan) [reply http-api/src/main/java/io/avaje/http/api/PathTypeConversion.java]: ok

File: http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java (hunk 2 above)
Comment (SentryMan) [reply http-generator-core/src/main/java/io/avaje/http/generator/core/ElementReader.java:90]: > `Optional<>` params are gonna haunt you with this.

Well, it's a good thing we don't support `Optional<>` then.

**Comment by SentryMan** [comment]
And I'm done, @rbygrave  plz review when you get time.

**Comment by SentryMan** [comment]
Split HttpClient changes into it's own PR #169

**Pull request merged**